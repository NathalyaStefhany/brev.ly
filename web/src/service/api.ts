import { env } from '@/env';
import axios from 'axios';

const baseURL = env.VITE_BACKEND_URL;

export const api = axios.create({
  baseURL,
});
