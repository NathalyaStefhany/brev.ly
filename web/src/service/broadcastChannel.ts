export const refetchChannel = new BroadcastChannel('refetch-channel');
