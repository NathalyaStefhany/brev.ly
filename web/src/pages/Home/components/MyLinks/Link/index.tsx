import React, { useState } from 'react';

import { Copy, Trash } from '@phosphor-icons/react';
import * as AlertDialog from '@radix-ui/react-alert-dialog';
import { IconButton } from '@/components/IconButton';
import { Button } from '@/components/Button';
import { api } from '@/service/api';
import { queryClient } from '@/service/queryClient';
import { Toast } from '@/components/Toast';

type LinkProps = {
  isFirstLink?: boolean;
  isLoading?: boolean;
  info?: {
    id: string;
    originalLink: string;
    shortenedLink: string;
    quantityAccesses: number;
  };
};

export const Link: React.FC<LinkProps> = ({
  isFirstLink = false,
  isLoading = false,
  info,
}) => {
  const [linkCopied, setLinkCopied] = useState(false);

  const [openDeleteDialog, setOpenDeleteDialog] = useState(false);
  const [isDeleting, setIsDeleting] = useState(false);
  const [deletionError, setDeletionError] = useState(false);

  const { id, originalLink, shortenedLink, quantityAccesses } = info ?? {};

  const handleCopyShortenedLink = (): void => {
    const link = `${window.location.origin}/${shortenedLink}`;

    navigator.clipboard.writeText(link);

    setLinkCopied(true);
  };

  const handleDeleteLink = async (): Promise<void> => {
    setIsDeleting(true);

    try {
      await api.delete(`/shortened-links/${shortenedLink}`);

      queryClient.refetchQueries({ queryKey: ['links list'], exact: true });

      setOpenDeleteDialog(false);
    } catch {
      setDeletionError(true);
    }

    setIsDeleting(false);
  };

  return (
    <>
      <div
        className="flex flex-row gap-8 md:gap-10 items-center border-t border-gray-200 pt-3 md:pt-4"
        style={{
          borderTopWidth: isFirstLink ? '0px' : '1px',
        }}
        data-testid={
          isLoading ? `container-loading-link` : `container-${id}-link`
        }
      >
        <div className="flex flex-col flex-1 gap-2 overflow-hidden">
          {isLoading ? (
            <div className="h-9 w-2/4 bg-gray-300 rounded animate-pulse" />
          ) : (
            <a
              href={`/${shortenedLink}`}
              target="_blank"
              rel="noreferrer"
              className="text-md text-blue-base truncate"
              data-testid="link-shortened-link"
            >
              brev.ly/{shortenedLink}
            </a>
          )}

          {isLoading ? (
            <div className="h-8 w-3/4 bg-gray-300 rounded animate-pulse" />
          ) : (
            <p
              className="text-sm text-gray-500 truncate"
              data-testid="text-original-link"
            >
              {originalLink}
            </p>
          )}
        </div>

        {isLoading ? (
          <div className="h-8 w-27 bg-gray-300 rounded animate-pulse" />
        ) : (
          <p
            className="text-sm text-gray-500 whitespace-nowrap"
            data-testid="text-quantity-accesses"
          >
            {quantityAccesses} acessos
          </p>
        )}

        <div className="flex flex-row gap-2 items-center">
          {isLoading ? (
            <>
              <div className="h-16 w-16 bg-gray-300 rounded animate-pulse" />{' '}
              <div className="h-16 w-16 bg-gray-300 rounded animate-pulse" />
            </>
          ) : (
            <>
              <IconButton
                icon={Copy}
                onClick={handleCopyShortenedLink}
                data-testid="button-copy"
              />

              <AlertDialog.Root
                open={openDeleteDialog}
                onOpenChange={setOpenDeleteDialog}
              >
                <AlertDialog.Overlay className="fixed inset-0 bg-gray-600/30 z-2" />

                <AlertDialog.Trigger asChild>
                  <IconButton icon={Trash} data-testid="button-delete" />
                </AlertDialog.Trigger>

                <AlertDialog.Content
                  className="w-150 z-3 fixed top-1/2 left-1/2 transform -translate-x-1/2 -translate-y-1/2 bg-white shadow-lg shadow-gray-400 p-8 rounded-lg"
                  data-testid="container-delete-confirmation"
                >
                  <AlertDialog.Title className="text-lg text-gray-600 mb-4">
                    Apagar link
                  </AlertDialog.Title>

                  <AlertDialog.Description className="text-sm text-gray-600 mb-12">
                    Você realmente quer apagar o link {shortenedLink}?
                  </AlertDialog.Description>

                  <div className="flex flex-row gap-4 justify-end">
                    <AlertDialog.Cancel asChild>
                      <Button
                        variant="secondary"
                        style={{ padding: '0.5rem 1.25rem' }}
                        data-testid="button-cancel-deletion"
                      >
                        Cancelar
                      </Button>
                    </AlertDialog.Cancel>

                    <Button
                      onClick={handleDeleteLink}
                      isLoading={isDeleting}
                      disabled={isDeleting}
                      style={{
                        width: 'fit-content',
                        padding: '0.5rem 1.25rem',
                        borderRadius: '4px',
                      }}
                      data-testid="button-delete-confirmation"
                    >
                      Apagar
                    </Button>
                  </div>
                </AlertDialog.Content>
              </AlertDialog.Root>
            </>
          )}
        </div>
      </div>

      <Toast
        id="toast-shortened-link-copied"
        type="information"
        title="Link copiado com sucesso"
        description={`O link ${shortenedLink} foi copiado para a área de transferência`}
        duration={3000}
        open={linkCopied}
        onOpenChange={setLinkCopied}
      />

      <Toast
        id="toast-deletion-error"
        type="error"
        title="Erro ao deletar"
        description="Por favor, tente novamente mais tarde."
        open={deletionError}
        onOpenChange={setDeletionError}
      />
    </>
  );
};
