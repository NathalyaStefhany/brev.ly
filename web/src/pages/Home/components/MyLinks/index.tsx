import React from 'react';

import {
  DownloadSimple,
  LinkBreak,
  Link as LinkIcon,
} from '@phosphor-icons/react';
import { useInfiniteQuery } from '@tanstack/react-query';
import * as ScrollArea from '@radix-ui/react-scroll-area';
import { Button } from '@/components/Button';
import { api } from '@/service/api';
import { Spinner } from '@/components/Spinner';
import { Link } from '@/pages/Home/components/MyLinks/Link';
import { EmptyState } from '@/components/EmptyState';

const PAGE_SIZE = 20;

type ShortenedLinkInfo = {
  id: string;
  originalLink: string;
  shortenedLink: string;
  quantityAccesses: number;
  createdAt: string;
};

type ShortenedLinkList = {
  total: number;
  page: number;
  pageSize: number;
  data: ShortenedLinkInfo[];
};

export const MyLinks: React.FC = () => {
  const {
    data: listOfLinks,
    isError: listOfLinksIsError,
    isFetching: listOfLinksIsFetching,
    isRefetching: listOfLinksIsRefetching,
    hasNextPage,
    fetchNextPage,
  } = useInfiniteQuery<ShortenedLinkList>({
    queryKey: ['links list'],
    queryFn: async ({ pageParam }) => {
      const { data } = await api.get(
        `/shortened-links?page=${pageParam}&pageSize=${PAGE_SIZE}`,
      );

      return data;
    },
    initialPageParam: 1,
    getNextPageParam: (lastPage, allPages) => {
      if (lastPage.total > allPages.length * PAGE_SIZE) {
        return lastPage.page + 1;
      }

      return undefined;
    },
  });

  const onScroll = (event: React.UIEvent<HTMLDivElement>): void => {
    const { scrollHeight, scrollTop, clientHeight } = event.currentTarget;

    const distanceFromScrollToEnd = scrollHeight - scrollTop - clientHeight;
    const scrollIsMiddle = distanceFromScrollToEnd <= clientHeight / 2;

    if (scrollIsMiddle && hasNextPage && !listOfLinksIsFetching) {
      fetchNextPage();
    }
  };

  const quantityOfLinks = (listOfLinks && listOfLinks.pages[0].total) ?? 0;
  const listOfLinksIsEmpty = quantityOfLinks === 0;

  return (
    <div className="relative max-w-190 w-full md:max-w-290 md:min-w-190 h-fit flex flex-col flex-1 gap-8 md:gap-10 bg-gray-100 rounded-lg p-12 md:p-16">
      {listOfLinksIsRefetching && (
        <div className="absolute top-0 left-0 h-1 w-full animate-border bg-size-[100px_auto] md:bg-size-[200px_auto] bg-no-repeat bg-gradient-to-r from-blue-base to-blue-base" />
      )}

      <div className="flex flex-row items-center justify-between">
        <h2 className="text-lg text-gray-600">Meus links</h2>

        <Button variant="secondary" icon={DownloadSimple} disabled>
          Baixar CSV
        </Button>
      </div>

      <div className="border-t border-gray-200">
        {listOfLinksIsError ? (
          <EmptyState
            icon={<LinkBreak size="2rem" color="var(--color-danger)" />}
            description="Erro ao carregar os links cadastrados"
            data-testid="container-link-list-error"
          />
        ) : !listOfLinks ? (
          <EmptyState
            icon={<Spinner />}
            description="Carregando..."
            data-testid="container-link-list-loading"
          />
        ) : listOfLinksIsEmpty ? (
          <EmptyState
            icon={<LinkIcon size="2rem" color="var(--color-gray-400)" />}
            description="Ainda nÃ£o existem links cadastrados"
            data-testid="container-link-list-empty"
          />
        ) : (
          <ScrollArea.Root type="auto" className="w-full">
            <ScrollArea.Viewport
              onScroll={onScroll}
              className="overflow-hidden"
              style={{
                maxHeight: 'calc(100vh - 19rem)',
                minHeight: quantityOfLinks < 4 ? 'fit-content' : '14.125rem',
              }}
              data-testid="container-my-links-scroll"
            >
              <div className="md:w-full flex flex-col gap-3 md:gap-4">
                {listOfLinks.pages.map(({ page, data }) =>
                  data.map((link, index) => {
                    const isFirstLink = index === 0 && page === 1;

                    return (
                      <Link
                        key={link.id}
                        isFirstLink={isFirstLink}
                        info={link}
                      />
                    );
                  }),
                )}

                {hasNextPage && <Link isFirstLink={false} isLoading />}
              </div>
            </ScrollArea.Viewport>

            <ScrollArea.Scrollbar
              orientation="vertical"
              className="flex w-4 mr-[-0.875rem] bg-gray-200 rounded-full transition-colors duration-150 ease-out"
            >
              <ScrollArea.Thumb className="flex-1 rounded-full bg-blue-base hover:bg-blue-dark" />
            </ScrollArea.Scrollbar>
          </ScrollArea.Root>
        )}
      </div>
    </div>
  );
};
