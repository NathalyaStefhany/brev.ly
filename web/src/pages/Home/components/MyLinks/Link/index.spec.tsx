import { beforeEach, describe, expect, it, vi } from 'vitest';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import AxiosMock from 'axios-mock-adapter';
import { Link } from '@/pages/Home/components/MyLinks/Link';
import { api } from '@/service/api';
import { queryClient } from '@/service/queryClient';

const apiMock = new AxiosMock(api);

const linkInfo = {
  id: '123',
  originalLink: 'https://test.com',
  shortenedLink: 'test',
  quantityAccesses: 0,
};

describe('Link tests', () => {
  beforeEach(() => {
    apiMock.reset();
  });

  it('should render correctly when it is first link', () => {
    const { container } = render(<Link isFirstLink info={linkInfo} />);

    expect(container).toMatchSnapshot();
  });

  it('should render correctly when it is not first link', () => {
    const { container } = render(<Link isFirstLink={false} info={linkInfo} />);

    expect(container).toMatchSnapshot();
  });

  it('should render correctly when it is loading', () => {
    const { container } = render(
      <Link isFirstLink info={linkInfo} isLoading />,
    );

    expect(container).toMatchSnapshot();
  });

  it('should diplay shortened link', () => {
    render(<Link isFirstLink={false} info={linkInfo} />);

    expect(screen.getByTestId('link-shortened-link')).toHaveTextContent(
      'brev.ly/test',
    );
  });

  it('should diplay original link', () => {
    render(<Link isFirstLink={false} info={linkInfo} />);

    expect(screen.getByTestId('text-original-link')).toHaveTextContent(
      'https://test.com',
    );
  });

  it('should diplay quantity of accesses', () => {
    render(<Link isFirstLink={false} info={linkInfo} />);

    expect(screen.getByTestId('text-quantity-accesses')).toHaveTextContent(
      '0 acessos',
    );
  });

  it('should copy shortened link', () => {
    const writeText = vi.fn();

    Object.assign(navigator, {
      clipboard: {
        writeText,
      },
    });

    render(<Link isFirstLink={false} info={linkInfo} />);

    fireEvent.click(screen.getByTestId('button-copy'));

    expect(writeText).toHaveBeenCalledWith('http://localhost:3000/test');

    expect(screen.getByTestId('toast-shortened-link-copied')).toHaveTextContent(
      'Link copiado com sucesso',
    );
    expect(screen.getByTestId('toast-shortened-link-copied')).toHaveTextContent(
      'O link test foi copiado para a área de transferência',
    );
  });

  it('should delete the link', async () => {
    apiMock.onDelete('/shortened-links/123').reply(200);

    queryClient.refetchQueries = vi.fn();

    render(<Link isFirstLink={false} info={linkInfo} />);

    expect(
      screen.queryByTestId('container-delete-confirmation'),
    ).not.toBeInTheDocument();

    fireEvent.click(screen.getByTestId('button-delete'));

    expect(
      screen.getByTestId('container-delete-confirmation'),
    ).toBeInTheDocument();

    expect(screen.queryByTestId('button-loading')).not.toBeInTheDocument();

    fireEvent.click(screen.getByTestId('button-delete-confirmation'));

    expect(screen.getByTestId('button-delete-confirmation')).toBeDisabled();
    expect(screen.getByTestId('button-loading')).toBeInTheDocument();

    await waitFor(() => {
      expect(
        apiMock.history.delete.some(
          ({ url }) => url === '/shortened-links/123',
        ),
      ).toBeTruthy();
    });

    expect(queryClient.refetchQueries).toBeCalledWith({
      queryKey: ['links list'],
      exact: true,
    });

    await waitFor(() => {
      expect(
        screen.queryByTestId('container-delete-confirmation'),
      ).not.toBeInTheDocument();
    });
  });
});
