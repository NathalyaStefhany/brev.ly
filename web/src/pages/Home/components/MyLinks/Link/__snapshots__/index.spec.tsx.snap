// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Link tests > should render correctly when it is first link 1`] = `
<div>
  <div
    class="flex flex-row gap-8 md:gap-10 items-center border-t border-gray-200 pt-3 md:pt-4"
    data-testid="container-123-link"
    style="border-top-width: 0px;"
  >
    <div
      class="flex flex-col flex-1 gap-2 overflow-hidden"
    >
      <a
        class="text-md text-blue-base truncate"
        data-testid="link-shortened-link"
        href="/test"
      >
        brev.ly/
        test
      </a>
      <p
        class="text-sm text-gray-500 truncate"
        data-testid="text-original-link"
      >
        https://test.com
      </p>
    </div>
    <p
      class="text-sm text-gray-500 whitespace-nowrap"
      data-testid="text-quantity-accesses"
    >
      0
       acessos
    </p>
    <div
      class="flex flex-row gap-2 items-center"
    >
      <button
        class="cursor-pointer transition-all duration-300 bg-gray-200 border border-gray-200 rounded-sm p-3.5 hover:not-disabled:border-blue-base disabled:opacity-50 disabled:cursor-not-allowed"
        type="button"
      >
        <svg
          fill="var(--color-gray-600)"
          height="1rem"
          viewBox="0 0 256 256"
          width="1rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M216,32H88a8,8,0,0,0-8,8V80H40a8,8,0,0,0-8,8V216a8,8,0,0,0,8,8H168a8,8,0,0,0,8-8V176h40a8,8,0,0,0,8-8V40A8,8,0,0,0,216,32ZM160,208H48V96H160Zm48-48H176V88a8,8,0,0,0-8-8H96V48H208Z"
          />
        </svg>
      </button>
      <button
        class="cursor-pointer transition-all duration-300 bg-gray-200 border border-gray-200 rounded-sm p-3.5 hover:not-disabled:border-blue-base disabled:opacity-50 disabled:cursor-not-allowed"
        type="button"
      >
        <svg
          fill="var(--color-gray-600)"
          height="1rem"
          viewBox="0 0 256 256"
          width="1rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M216,48H176V40a24,24,0,0,0-24-24H104A24,24,0,0,0,80,40v8H40a8,8,0,0,0,0,16h8V208a16,16,0,0,0,16,16H192a16,16,0,0,0,16-16V64h8a8,8,0,0,0,0-16ZM96,40a8,8,0,0,1,8-8h48a8,8,0,0,1,8,8v8H96Zm96,168H64V64H192ZM112,104v64a8,8,0,0,1-16,0V104a8,8,0,0,1,16,0Zm48,0v64a8,8,0,0,1-16,0V104a8,8,0,0,1,16,0Z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`Link tests > should render correctly when it is loading 1`] = `
<div>
  <div
    class="flex flex-row gap-8 md:gap-10 items-center border-t border-gray-200 pt-3 md:pt-4"
    data-testid="container-loading-link"
    style="border-top-width: 0px;"
  >
    <div
      class="flex flex-col flex-1 gap-2 overflow-hidden"
    >
      <div
        class="h-9 w-2/4 bg-gray-300 rounded animate-pulse"
      />
      <div
        class="h-8 w-3/4 bg-gray-300 rounded animate-pulse"
      />
    </div>
    <div
      class="h-8 w-27 bg-gray-300 rounded animate-pulse"
    />
    <div
      class="flex flex-row gap-2 items-center"
    >
      <div
        class="h-16 w-16 bg-gray-300 rounded animate-pulse"
      />
       
      <div
        class="h-16 w-16 bg-gray-300 rounded animate-pulse"
      />
    </div>
  </div>
</div>
`;

exports[`Link tests > should render correctly when it is not first link 1`] = `
<div>
  <div
    class="flex flex-row gap-8 md:gap-10 items-center border-t border-gray-200 pt-3 md:pt-4"
    data-testid="container-123-link"
    style="border-top-width: 1px;"
  >
    <div
      class="flex flex-col flex-1 gap-2 overflow-hidden"
    >
      <a
        class="text-md text-blue-base truncate"
        data-testid="link-shortened-link"
        href="/test"
      >
        brev.ly/
        test
      </a>
      <p
        class="text-sm text-gray-500 truncate"
        data-testid="text-original-link"
      >
        https://test.com
      </p>
    </div>
    <p
      class="text-sm text-gray-500 whitespace-nowrap"
      data-testid="text-quantity-accesses"
    >
      0
       acessos
    </p>
    <div
      class="flex flex-row gap-2 items-center"
    >
      <button
        class="cursor-pointer transition-all duration-300 bg-gray-200 border border-gray-200 rounded-sm p-3.5 hover:not-disabled:border-blue-base disabled:opacity-50 disabled:cursor-not-allowed"
        type="button"
      >
        <svg
          fill="var(--color-gray-600)"
          height="1rem"
          viewBox="0 0 256 256"
          width="1rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M216,32H88a8,8,0,0,0-8,8V80H40a8,8,0,0,0-8,8V216a8,8,0,0,0,8,8H168a8,8,0,0,0,8-8V176h40a8,8,0,0,0,8-8V40A8,8,0,0,0,216,32ZM160,208H48V96H160Zm48-48H176V88a8,8,0,0,0-8-8H96V48H208Z"
          />
        </svg>
      </button>
      <button
        class="cursor-pointer transition-all duration-300 bg-gray-200 border border-gray-200 rounded-sm p-3.5 hover:not-disabled:border-blue-base disabled:opacity-50 disabled:cursor-not-allowed"
        type="button"
      >
        <svg
          fill="var(--color-gray-600)"
          height="1rem"
          viewBox="0 0 256 256"
          width="1rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M216,48H176V40a24,24,0,0,0-24-24H104A24,24,0,0,0,80,40v8H40a8,8,0,0,0,0,16h8V208a16,16,0,0,0,16,16H192a16,16,0,0,0,16-16V64h8a8,8,0,0,0,0-16ZM96,40a8,8,0,0,1,8-8h48a8,8,0,0,1,8,8v8H96Zm96,168H64V64H192ZM112,104v64a8,8,0,0,1-16,0V104a8,8,0,0,1,16,0Zm48,0v64a8,8,0,0,1-16,0V104a8,8,0,0,1,16,0Z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
